UTM support for `view --show-gps` + dual-coord input for `lalo2yx/yx2lalo()`

forrestfwilliams opened the pull request:
**Description of proposed changes**

HyP3 (and I believe NISAR as well) both output their InSAR products in UTM coordinates. While MintPy was originally designed to use lat/lon coordinates, has limited support for UTM datasets as well. Users have noted (i.e., issue #1145) that it would be nice to better support UTM datasets.

This PR adds some additional support for UTM datasets, in plotting and GPS comparison contexts by:
1. Ensuring UTM coordinate plot axes do not use scientific notation (see issue #1145)
2. Ensuring `coords.coordinate.latlon2yx` works with UTM coordinates
3. Ensuring `coords.coordinate.geo2rdr` works with UTM coordinates
4. Adding compatibility for UTM coordinates to GPS plotting functionality

To accomplish this, checks for the use of UTM coordinates were added to `latlon2yx` and `geo2rdr`. Also, the inputs for `latlon2yx` was changed from a coordinate and a coordinate type (lat or lon) to a pair of lat/lon or northing/easting coordinates. All uses of `latlon2yx` were changed to use this new input format.

**Reminders**

- [x] Fixes #1145
- [x] Pass Pre-commit check (green)
- [x] Pass Codacy code review (green)
- [x] Pass Circle CI test (green)
- [x] Make sure that your code follows our style. Use the other functions/files as a basis.
- [x] If modifying functionality, describe changes to function behavior and arguments in a comment below the function declaration.
- [x] If adding new functionality, add a detailed description to the documentation and/or an example.

forrestfwilliams commented:
Also related to discussion #1010

forrestfwilliams commented:
Hey @yunjunz, I would appreciate your thoughts on this whenever you have time.

yunjunz commented on src/mintpy/utils/plot.py:
```diff
@@ -1214,6 +1220,9 @@ def plot_gps(ax, SNWE, inps, metadata=dict(), print_msg=True):
             vprint(f'ignore the following {np.sum(nan_flag)} stations due to limited overlap/observations in time')
             vprint(f'  {site_names[nan_flag]}')
 
+        if 'UTM_ZONE' in metadata:
+            site_lats, site_lons = ut0.latlon2utm(site_lats, site_lons)
+
```
My local testing based on the Ridgecrest example dataset as below did not display it right (all GPS are concentrated to the origin near 0), and moving this part of the code [here](https://github.com/insarlab/MintPy/blob/7ac345551ac237f0e1a314cd3f04593f4ea1305f/src/mintpy/utils/plot.py#L1136) fixed it.

```bash
view.py velocity.h5 velocity --show-gps
```

yunjunz commented on src/mintpy/objects/coord.py:
```diff
@@ -68,44 +68,50 @@ def open(self):
+        lat_coord_in = self._clean_coord(lat_coord_in)
+        lon_coord_in = self._clean_coord(lon_coord_in)
+
+        if 'UTM_ZONE' in self.src_metadata and np.max(lat_coord_in) <= 90 and np.max(lon_coord_in) <= 360:
```
```suggestion
        if 'UTM_ZONE' in self.src_metadata and np.max(np.abs(lat_coord_in)) <= 90 and np.max(np.abs(lon_coord_in)) <= 360:
```

yunjunz commented on src/mintpy/objects/coord.py:
```diff
@@ -214,11 +220,15 @@ def _read_geo_lut_metadata(self):
                     az/rg_res - float, residul/uncertainty of coordinate conversion
         """
         # ensure longitude convention to be consistent with src_metadata
+        if 'UTM_ZONE' in self.src_metadata and (np.max(lat) > 90 or np.max(lon) > 360):
```
```suggestion
        if 'UTM_ZONE' in self.src_metadata and (np.max(np.abs(lat)) > 90 or np.max(np.abs(lon)) > 360):
```

yunjunz commented on src/mintpy/objects/coord.py:
```diff
@@ -68,44 +68,50 @@ def open(self):
         if isinstance(coord_in, np.ndarray):
             coord_in = coord_in.tolist()
         # note: np.float128 is not supported on Windows OS, use np.longdouble as a platform neutral syntax
         if isinstance(coord_in, (float, np.float16, np.float32, np.float64, np.longdouble)):
```
If we go with the pair style, let's add the integer type here as well, so that we can reuse this func in `yx2lalo()`.

forrestfwilliams replied:
done!

yunjunz reviewed (commented):
Thank you @forrestfwilliams for the PR! After some minor changes mentioned above, testing the GPS plotting via view.py worked for me. 

As for the input arg changes for `lalo2yx()`, either the "single coordinate type" or the "pair style" is fine with me. Actually, the pair style looks cleaner. Could you do a similar change to `yx2lalo()` as well, for consistency?

forrestfwilliams commented:
Thanks for the review @yunjunz! I've made the suggested changes, modified `yx2lalo` to match the structure of `lalo2yx`, and updated its uses. This is ready for you to look at again!

forrestfwilliams commented:
Not sure why the pre-commit check is failing.

yunjunz reviewed (approved):
Looks all good to me now. Thank you @forrestfwilliams!

Although `lalo2yx()` and `yx2lalo()` have breaking changes, they are usually not heavily used by most developers I believe, as `geo2radar()` and `radar2geo()` are more generic and convenient, I think there will be only very limited impact on developers and advanced users.

forrestfwilliams commented:
Awesome, thanks @yunjunz! Do you think we could create a new release of MintPy soon with these changes? I ended up pursuing this work to support some tutorials ASF's OpenSARLab team runs and it would be great to get them a new release with these changes.

yunjunz commented:
Sure thing, let's cut for a new release for it. There are a few recently open PRs from the JPL folks, it will be nice to merge those before the new release, likely this week.

yunjunz commented:
Hi @forrestfwilliams, I think the code is ready for a new release. As for the proposed changes in https://github.com/insarlab/MintPy/issues/1162, doing them either before or after the release is fine with me, let me know please.

forrestfwilliams commented:
Hey @yunjunz sorry for the slow response. I'll open a PR for the changes discussed in #1162 next week, then we can release!
